File: common/src/model/login_info.rs
use crate::model::cookies::CookieItem;
use crate::model::headers::HeaderItem;
use crate::model::{Cookies, Headers};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LoginInfo {
    pub cookies: Vec<CookieItem>,
    pub useragent: String,
    pub extra: serde_json::Value,
}

impl Default for LoginInfo {
    fn default() -> Self {
        LoginInfo{
            cookies: vec![],
            useragent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/139.0.0.0 Safari/537.36".to_string(),
            extra: Default::default(),
        }
    }
}

impl From<&LoginInfo> for Cookies {
    fn from(value: &LoginInfo) -> Self {
        Cookies {
            cookies: value.cookies.clone(),
        }
    }
}

impl From<&LoginInfo> for Headers {
    fn from(value: &LoginInfo) -> Self {
        Headers {
            headers: vec![HeaderItem {
                key: "User-Agent".to_string(),
                value: value.useragent.clone(),
            }],
        }
    }
}
impl LoginInfo {
    pub fn get_extra<T>(&self, key: &str) -> Option<T>
    where
        T: Serialize + for<'de> Deserialize<'de>,
    {
        self.extra
            .get(key)
            .and_then(|v| serde_json::from_value::<T>(v.clone()).ok())
    }
    pub fn get_shop_id(&self) -> Option<i64> {
        self.get_extra("shopid")
    }
}

File: python_mocra/common/models/login_info.py
from typing import List, Optional, Any, Dict
from pydantic import BaseModel, Field
from .cookies import CookieItem, Cookies
from .headers import Headers, HeaderItem

class LoginInfo(BaseModel):
    cookies: List[CookieItem] = Field(default_factory=list)
    useragent: str = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/139.0.0.0 Safari/537.36"
    extra: Dict[str, Any] = Field(default_factory=dict)

    def to_cookies(self) -> Cookies:
        return Cookies(cookies=self.cookies)

    def to_headers(self) -> Headers:
        return Headers(headers=[HeaderItem(key="User-Agent", value=self.useragent)])

    def get_extra(self, key: str) -> Optional[Any]:
        return self.extra.get(key)
    
    def get_shop_id(self) -> Optional[int]:
        return self.extra.get("shopid")
